src/core/server/saved_objects/migrationsv2/index.ts

```diff
@@ -0,0 +1,240 @@
+        res.fetchAliases['.kibana_7.11.0'] != null &&
+        res.fetchAliases['.kibana_current'] === res.fetchAliases['.kibana_7.11.0']
+      ) {
+        // `.kibana_current` and the version specific aliases both exists and
+        // are pointing to the same index. This version's migration has already
+        // been completed.
+        stateP.controlState = 'DONE';
+        // TODO, go to step (6)
+      } else if (
+        res.fetchAliases['.kibana_current'] != null &&
+        gt(indexVersion(res.fetchAliases['.kibana_current']), '7.11.0')
+      ) {
+        // `.kibana_current` is pointing to an index that belongs to a later
+        // version of Kibana .e.g. a 7.11.0 node found `.kibana_7.12.0_001`
+        stateP.controlState = 'FATAL';
+        stateP.error = new Error(
+          'The .kibana_current alias is pointing to a newer version of Kibana: v' +
+            indexVersion(res.fetchAliases['.kibana_current'])
+        );
+      } else if (res.fetchAliases['.kibana_current'] ?? res.fetchAliases['.kibana'] ?? false) {
+        //  The source index is:
+        //  1. the index the `.kibana_current` alias points to, or if it doesn’t exist,
+        //  2. the index the `.kibana` alias points to, or if it doesn't exist,
+        const source = res.fetchAliases['.kibana_current'] || res.fetchAliases['.kibana'];
+
+        stateP.controlState = 'SET_SOURCE_WRITE_BLOCK';
+        stateP.source = source;
+        stateP.target = `.kibana_${stateP.kibanaVersion}_001`;
+      } else {
+        // TODO: Check if we need to migrate from a v6.x `.kibana` index, source = '.kibana'
+        // This cluster doesn't have any existing Kibana indices, create a new
+        // version specific index.
+        stateP.controlState = 'INIT_NEW_INDICES';
+        stateP.target = `.kibana_${stateP.kibanaVersion}_001`;
+      }
+    }
+    return stateP;
+  } else if (control === 'SET_SOURCE_WRITE_BLOCK') {
+    if ('setIndexWriteBlock' in res) {
+      stateP.controlState = 'CLONE_SOURCE';
+    }
+    return stateP;
+  } else if (control === 'CLONE_SOURCE') {
+    if ('cloneIndex' in res) {
+      stateP.controlState = 'DONE';
+    }
+    return stateP;
+  } else if (control === 'DONE' || control === 'FATAL' || control === 'INIT_NEW_INDICES') {
+    return stateP;
+  } else {
+    return throwBadControlState(control);
+  }
+};
+
+export const next: NextAction = (client: ElasticsearchClient, state: State) => {
+  // If for every state there is only one action (deterministic) we can define
+  // next as a state -> action set.
+  const delay = <F extends () => any>(fn: F): (() => ReturnType<F>) => {
+    return () => {
+      return state.retryDelay > 0
+        ? new Promise((resolve) => setTimeout(resolve, state.retryDelay)).then(fn)
+        : fn();
+    };
+  };
+  const control = state.controlState;
+  if (control === 'INIT') {
+    return delay(Actions.fetchAliases(client, ['.kibana', '.kibana_current', '.kibana_7.11.0']));
```

Comment:
nit: the action is tightly coupled with the state. Is it possible to collocate them closer?